keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,649 @@
+
+#### No support JobSuccessPolicy for the NonIndexed Job
+
+The JobSuccessPolicy can be theoretically supported for the NonIndexed job.
+However, the JobSuccessPolicy isn't supported for the NonIndexed job in the first iteration 
+since there aren't any effective use cases for the NonIndexed job.
+
+#### Interchangeability between concepts of `completed` and `succeeded`
+In this KEP, we introduce the new `succeeded` concept, 
+which is interchangeable with the existing `completed` concept.
+
+It means that if the job meets the dedicated successPolicy, the `Completed` condition is added to the job.  
+
+#### Surprising whenTargetAchieved behavior if whenTargetAchieved is not specified somewhere  
+
+When the successPolicy has 2 targets and the first target doesn't have `whenTargetAchieved`,
+a `whenTargetAchieved=terminate` is set by default.  
+
+```yaml
+...
+spec:
+  successPolicy:
+    - target:
+        succeededIndexes: "0"
+    - target:
+        succeededIndexes: "1-2"
+        whenTargetAchieved: continue
+```
+
+In that case, if all targets are achieved simultaneously, the first target is evaluated, and all pods are terminated at the first.
+As a result, the second target is not evaluated, and the second `whenTargetAchieved` value is ignored.
+
+This might be surprising behavior for users because the first no specified `whenTargetWhenAchieved` is performed,
+and the second specified `whenTargetWhenAchieved` is not performed.
+
+### Risks and Mitigations
+
+- If the job object's size reaches to limit of the etcd and
+the job controller can't store a correct value in `.status.completedIndexes`,
+we probably can not evaluate the SuccessPolicy correctly.
+
+- If we allow to set unlimited size of the value in `.spec.successPolicy.target.succeededIndexes`,
+we have a risk similar to [KEP-3850: Backoff Limits Per Index For Indexed Jobs](https://github.com/kubernetes/enhancements/tree/76dcd4f342cc0388feb085e685d4cc018ebe1dc9/keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs#risks-and-mitigations).
+So, we limit the size of `succeededIndexes` to 64Ki.
+
+## Design Details
+
+### Job API
+
+We extend the Job API to set different policies by which a Job can be declared as succeeded.
+
+```golang
+type JobSpec struct {
+	...
+	// Specifies the policy when the job can be declared as succeeded.
+	// If empty, the default behavior applies - the job is declared as succeeded
+	// only when the number of succeeded pods equals to the completions.
+	// These policies are evaluated in order.
+	//
+	// This field is alpha-level. To use this field, you must enable the
+	// `JobSuccessPolicy` feature gate (disable by default).
+	// +optional
+    // +listType=atomic
+	SuccessPolicy []SuccessPolicy
+}
+
+// SuccessPolicy describes when a Job can be declared as succeeded based on the success of some indexes.
+type SuccessPolicy struct {
+	// Represents the requirement on the succeeded pods.
+	// +optional
+	Target *SuccessPolicyTarget
```

Comment:
As in discussed in the other threads, we may need to have mutliple targets in the future, connected by OR.

Maybe this is a place where we could consider having a list of targets / rules already, because I don't think we will be able to easily move the singleton to a list in the future. 

Even if there is no clear use case for now maybe this could be justified by the foreseeable need for extensibility? WDYT @tenzen-y @deads2k @alculquicondor ?

Reply:
> As in discussed in the other threads, we may need to have mutliple targets in the future, connected by OR.

@mimowo For more clarification, does that mean the following?

```yaml
...
  successPolicy:
  - targets:
    - succeededIndexes: "0"
      whenTargetAchieved: terminate
...
```

Current my proposal:

```yaml
...
  successPolicy:
  - target:
      succeededIndexes: "0"
      whenTargetAchieved: terminate
...
```

> Maybe this is a place where we could consider having a list of targets / rules already, because I don't think we will be able to easily move the singleton to a list in the future.
> 
> Even if there is no clear use case for now maybe this could be justified by the foreseeable need for extensibility?

It sounds reasonable to me, but I would like to know what @deads2k @alculquicondor's concerns are and what we should provide extensibility.

Reply:
I discussed this API offline with @mimowo and @deads2k.
As a result of that discussion, we agree with using the following API scheme so that we can have flexibility for the future:

```yaml
...
  successPolicy:
    criteria:
    - succeededIndexes: "0"
      whenTargetAchieved: terminate
    - succeededIndexes: "1-9"
      whenTargetAchieved: terminate
...
```

Reply:
Could that be mutated/changed?

Reply:
> Could that be mutated/changed?

No, it couldn't as well as podFailurePolicy. I don't think that we should make successPolicy mutable.

Reply:
+1 to the suggestion above, as  its more consistent with the existing pod failure policy:

```yaml
podFailurePolicy:
    rules:
    - action: FailJob
      onExitCodes:
        containerName: main
        operator: In
        values: [42]
```

Reply:
Done.

Reply:
Done.